### PR TITLE
enable mypy

### PR DESCRIPTION
**Pull request by trim21**
enable mypy on beancount

1. Add `mypy` and `types-regex` to `types` extra `project.optional-dependencies`
2. add mypy ci job to run mypy checks

this PR add mypy options `check_untyped_defs = false` so it only check function body with arguments/return type checks. untyped function like `def test_a(self):` will not be checked.

use `uv` in ci job there is no way to only install deps without building project from pyproject.toml with pip.

### CODE REVIEW

File: beancount/core/getters.py [hunk 1]
```diff
@@ -32,7 +34,7 @@
     from beancount.core.data import Note
     from beancount.core.data import Pad
 
-    AccountsUseMap = tuple[dict[Account, datetime.date], dict[Account, datetime.date]]
+    AccountsUseMap = Tuple[Dict[Account, datetime.date], Dict[Account, datetime.date]]
```
Comment (blais) [inline beancount/core/getters.py:37]: Why not keep the new modern types?

**Review by blais — APPROVED** [review]
LGTM (except the use of the older types)

File: beancount/core/getters.py (hunk 1 above)
Comment (blais) [reply beancount/core/getters.py:37]: I'll assume there's a mypy related reason; I'll go ahead and merge this, we can separately modernize all the type annotations after (there are plenty others).

Comment (trim21) [reply beancount/core/getters.py:37]: 1. we target py38 
2. in mypy `typing.TYPE_CHECKING` is always true, so mypy take it as runable code

Comment (blais) [reply beancount/core/getters.py:37]: Let's bump it up to 3.9 and require modern type annotations.
(In general I feel in branch `master` it's fine for us to require more modern Python. 
Even 3.9 is old by now, if 3.12 had something we care about I'd bump the req.)

Comment (blais) [reply beancount/core/getters.py:37]: Oh I wonder if gemini can modernize all the type annotations for me... I'm on an Aider kick...

Comment (trim21) [reply beancount/core/getters.py:37]: I'm still using py310...

Comment (blais) [reply beancount/core/getters.py:37]: AFAIK generics in standard type annotations became possible in 3.9, so you should be alright:
https://peps.python.org/pep-0585/

**Pull request merged**